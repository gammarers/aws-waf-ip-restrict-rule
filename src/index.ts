import * as wafv2 from 'aws-cdk-lib/aws-wafv2';

export interface RuleConfig {
  readonly priority: number;
  readonly ruleName?: string;
  readonly cloudWatchMetricsName?: string;
}

export interface WAFIPRestrictRuleProps {
  readonly allowIPSetArn: string;
}

export class WAFIPRestrictRule {

  constructor(private props: WAFIPRestrictRuleProps) {
  }

  allowRule(config: RuleConfig): wafv2.CfnWebACL.RuleProperty {
    return {
      name: config.ruleName || 'allow-ip-rule',
      priority: config.priority,
      action: { allow: {} },
      visibilityConfig: {
        cloudWatchMetricsEnabled: true,
        metricName: config.cloudWatchMetricsName || 'AllowIPMetric',
        sampledRequestsEnabled: true,
      },
      statement: {
        ipSetReferenceStatement: {
          arn: this.props.allowIPSetArn,
        },
      },
    };
  }

  blockRule(config: RuleConfig): wafv2.CfnWebACL.RuleProperty {
    return {
      name: config.ruleName || 'block-other-ip-rule',
      priority: 2,
      action: { block: {} }, // 拒否アクション
      visibilityConfig: {
        cloudWatchMetricsEnabled: true,
        metricName: config.cloudWatchMetricsName || 'BlockOtherIpMetric',
        sampledRequestsEnabled: true,
      },
      statement: {
        notStatement: {
          statement: {
            ipSetReferenceStatement: {
              arn: this.props.allowIPSetArn,
            },
          },
        },
      },
    };
  }
}
