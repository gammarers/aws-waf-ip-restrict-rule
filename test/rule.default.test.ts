import { App, Stack } from 'aws-cdk-lib';
import { Match, Template } from 'aws-cdk-lib/assertions';
import * as wafv2 from 'aws-cdk-lib/aws-wafv2';
import { WAFIPRestrictRule } from '../src';

describe('Web ACL Rule Global Cope testing', () => {

  const app = new App();
  const stack = new Stack(app, 'TestingStack');

  const allowedIpSet = new wafv2.CfnIPSet(stack, 'AllowedIpSet', {
    addresses: [
      '203.0.113.0/24',
      '198.51.100.0/24',
    ],
    ipAddressVersion: 'IPV4',
    scope: 'REGIONAL',
    name: 'AllowedIpSet',
  });

  const ipRestrictRule = new WAFIPRestrictRule({
    allowIPSetArn: allowedIpSet.attrArn,
  });

  new wafv2.CfnWebACL(stack, 'WebACL', {
    defaultAction: { allow: {} },
    scope: 'CLOUD_FRONT',
    name: 'WebAclWithCustomRules',
    visibilityConfig: {
      cloudWatchMetricsEnabled: true,
      metricName: 'WebAclMetric',
      sampledRequestsEnabled: true,
    },
    rules: [
      ipRestrictRule.allowRule({ priority: 1 }),
      ipRestrictRule.blockRule({ priority: 2 }),
    ],
  });

  it('Should have WAF WebACL Rule', async () => {
    template.hasResourceProperties('AWS::WAFv2::WebACL', {
      Rules: [
        {
          Action: {
            Allow: {},
          },
          Name: 'allow-ip-rule',
          Priority: 1,
          Statement: {
            IPSetReferenceStatement: {
              Arn: {
                'Fn::GetAtt': [
                  Match.stringLikeRegexp('AllowedIpSet'),
                  'Arn',
                ],
              },
            },
          },
          VisibilityConfig: {
            CloudWatchMetricsEnabled: true,
            MetricName: 'AllowIPMetric',
            SampledRequestsEnabled: true,
          },
        },
        {
          Action: {
            Block: {},
          },
          Name: 'block-other-ip-rule',
          Priority: 2,
          Statement: {
            NotStatement: {
              Statement: {
                IPSetReferenceStatement: {
                  Arn: {
                    'Fn::GetAtt': [
                      Match.stringLikeRegexp('AllowedIpSet'),
                      'Arn',
                    ],
                  },
                },
              },
            },
          },
          VisibilityConfig: {
            CloudWatchMetricsEnabled: true,
            MetricName: 'BlockOtherIpMetric',
            SampledRequestsEnabled: true,
          },
        },
      ],
    });
  });

  const template = Template.fromStack(stack);

  it('Should match snapshot', async () => {
    expect(template.toJSON()).toMatchSnapshot();
  });

});