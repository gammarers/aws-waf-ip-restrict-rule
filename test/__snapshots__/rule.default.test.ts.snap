// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Web ACL Rule Global Cope testing Should match snapshot 1`] = `
{
  "Parameters": {
    "BootstrapVersion": {
      "Default": "/cdk-bootstrap/hnb659fds/version",
      "Description": "Version of the CDK Bootstrap resources in this environment, automatically retrieved from SSM Parameter Store. [cdk:skip]",
      "Type": "AWS::SSM::Parameter::Value<String>",
    },
  },
  "Resources": {
    "AllowedIpSet": {
      "Properties": {
        "Addresses": [
          "203.0.113.0/24",
          "198.51.100.0/24",
        ],
        "IPAddressVersion": "IPV4",
        "Name": "AllowedIpSet",
        "Scope": "REGIONAL",
      },
      "Type": "AWS::WAFv2::IPSet",
    },
    "WebACL": {
      "Properties": {
        "DefaultAction": {
          "Allow": {},
        },
        "Name": "WebAclWithCustomRules",
        "Rules": [
          {
            "Action": {
              "Allow": {},
            },
            "Name": "allow-ip-rule",
            "Priority": 1,
            "Statement": {
              "IPSetReferenceStatement": {
                "Arn": {
                  "Fn::GetAtt": [
                    "AllowedIpSet",
                    "Arn",
                  ],
                },
              },
            },
            "VisibilityConfig": {
              "CloudWatchMetricsEnabled": true,
              "MetricName": "AllowIPMetric",
              "SampledRequestsEnabled": true,
            },
          },
          {
            "Action": {
              "Block": {},
            },
            "Name": "block-other-ip-rule",
            "Priority": 2,
            "Statement": {
              "NotStatement": {
                "Statement": {
                  "IPSetReferenceStatement": {
                    "Arn": {
                      "Fn::GetAtt": [
                        "AllowedIpSet",
                        "Arn",
                      ],
                    },
                  },
                },
              },
            },
            "VisibilityConfig": {
              "CloudWatchMetricsEnabled": true,
              "MetricName": "BlockOtherIpMetric",
              "SampledRequestsEnabled": true,
            },
          },
        ],
        "Scope": "CLOUD_FRONT",
        "VisibilityConfig": {
          "CloudWatchMetricsEnabled": true,
          "MetricName": "WebAclMetric",
          "SampledRequestsEnabled": true,
        },
      },
      "Type": "AWS::WAFv2::WebACL",
    },
  },
  "Rules": {
    "CheckBootstrapVersion": {
      "Assertions": [
        {
          "Assert": {
            "Fn::Not": [
              {
                "Fn::Contains": [
                  [
                    "1",
                    "2",
                    "3",
                    "4",
                    "5",
                  ],
                  {
                    "Ref": "BootstrapVersion",
                  },
                ],
              },
            ],
          },
          "AssertDescription": "CDK bootstrap stack version 6 required. Please run 'cdk bootstrap' with a recent version of the CDK CLI.",
        },
      ],
    },
  },
}
`;
